### DISCUSSION TITLE
Fedora impact check with 7.0.0rc1

### DISCUSSION THREAD

**Discussion by hroncok** [pytest-dev/pytest#9415] [Show and tell]
Hello, I am slowly triaging [the impact of pytest 7.0.0rc1 on Fedora packages](https://src.fedoraproject.org/rpms/pytest/pull-request/25#comment-92228). Will cross post upstream issues here.

See https://github.com/pytest-dev/pytest/discussions/9415#discussioncomment-1834432 for a full list of cross-references.

**Comment by bluetech** [comment]
@hroncok Great as always!

I will try to help with triage/fixes over the weekend.

Right now I only clicked `bandit` one: says `ModuleNotFoundError: No module named 'toml'` - it might be caused by pytest switching from `toml` to `tomli` and now the dep is missing. Probably needs an extra `toml` test dep (I do see it present in the test-requirements.txt file so maybe just missing in Fedora package).

**Reply by hroncok** [reply]
bandit in Fedora does not seem to install from https://github.com/PyCQA/bandit/blob/1.7.1/test-requirements.txt

**Reply by hroncok** [reply]
https://src.fedoraproject.org/rpms/bandit/pull-request/2

**Comment by The-Compiler** [comment]
# Initial triage

**Note: This post will be edited once the mentioned issues are "dealt with", i.e. the issues are triaged.**

From our perspective, that means that there is either:

- We opened an issue in the pytest repository for it
- or the issue has been reported upstream (ideally via a PR!) AND we're confident there is nothing pytest needs to do
- or we deem the issue a "wontfix" (e.g. dead upstream) AND see above

I took a quick look through all the logs, just to get a quick overview. Here is what I found:

## Packaging issues (Fedora specific)

Missing `toml` dependency (got pulled in via pytest before, which uses `tomli` now)

- [x] [bandit](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058752-bandit/builder-live.log.gz): fixed via https://src.fedoraproject.org/rpms/bandit/pull-request/2
- [x] [python-box](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058509-python-box/builder-live.log.gz): https://src.fedoraproject.org/rpms/python-box/pull-request/1
- [x] [python-vulture](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058532-python-vulture/builder-live.log.gz): https://src.fedoraproject.org/rpms/python-vulture/pull-request/1
- [x] [python-pytest-rerunfailures](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058527-python-pytest-rerunfailures/builder-live.log.gz): `E       AttributeError: 'list' object has no attribute 'items'` in `for col, (finalizers, exc) in self.stack.items():` <br> Can't seem to consistently reproduce - I was bisecting it and suddenly it's gone even with pytest 7 now<br>Already fixed in https://github.com/pytest-dev/pytest-rerunfailures/pull/151, which was [released](https://github.com/pytest-dev/pytest-rerunfailures/blob/master/CHANGES.rst) in pytest-rerunfailures 10.0 back in May, but Fedora still seems to be at 9.1.1 

## Fatal warnings or wrong usage

- [x] [python-diff-cover](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058512-python-diff-cover/builder-live.log.gz): `approx` in boolean context, fixed via https://github.com/Bachmann1234/diff_cover/pull/264
- [x] [python-ratelimiter](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058528-python-ratelimiter/builder-live.log.gz): `E   pytest.PytestRemovedIn7Warning: pytest.collect.File was moved to pytest.File` - **probably not worth trying to fix, it's related to code for Python < 3.5 and no upstream commits in over 4 years despite open issues/PRs**
- [x] [python-httplib2](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058516-python-httplib2/builder-live.log.gz): `Skip.__init__() got multiple values for argument 'reason' - maybe you meant pytest.mark.skipif?` - wrong API usage: https://github.com/httplib2/httplib2/pull/215

Various `warns(None)` usage, see https://github.com/pytest-dev/pytest/issues/9002:

- [x] [python-aiohttp](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7:pr:2/fedora-rawhide-x86_64/03084226-python-aiohttp/builder-live.log.gz): `warns(None)`
- [x] [python-trio](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058531-python-trio/builder-live.log.gz): `warns(None)`
- [x] [python-flask](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058511-python-flask/builder-live.log.gz): `warns(None)`
- [x] [python-ase](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058508-python-ase/builder-live.log.gz): `warns(None)`
- [x] [python-dateutil](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058510-python-dateutil/builder-live.log.gz): `warns(None)`

## Private API usage

- [x] [pg_activity](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058504-pg_activity/builder-live.log.gz): `from _pytest.tmpdir import TempdirFactory` via pytest_postgresql
- [x] [python-astroscrappy](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058506-python-astroscrappy/builder-live.log.gz): `E   TypeError: Config._getconftest_pathlist() missing 1 required positional argument: 'rootpath'` (`collect_ignore = config._getconftest_pathlist("collect_ignore", path=dirpath)`) <br>Happened via [`pytest-doctestplus`](https://github.com/astropy/pytest-doctestplus), already fixed in its [latest releases](https://github.com/astropy/pytest-doctestplus/releases) and newer Fedora tests [seem to succeed too](https://copr.fedorainfracloud.org/coprs/churchyard/pytest-7/package/python-astroscrappy/)
- [x] [black](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058501-python-black/builder-live.log.gz): `ImportError: Error importing plugin "tests.optional": No module named '_pytest.store'` (reported upstream: https://github.com/psf/black/pull/2705)
- [x] [python-pytest-cases](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058524-python-pytest-cases/builder-live.log.gz): `from _pytest.fixtures import scopes`: https://github.com/smarie/python-pytest-cases/issues/241
- [x] [python-fiat](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058514-python-fiat/builder-live.log.gz): `from _pytest.python import scopes` via `pytest-cases`
- [x] [python-pytest-postgresql](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058525-python-pytest-postgresql/builder-live.log.gz): `from _pytest.tmpdir import TempdirFactory`: https://github.com/ClearcodeHQ/pytest-postgresql/issues/533

## Incorrect version parsing

- [x] [python-mirrors-countme](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058518-python-mirrors-countme/builder-live.log.gz): `E   ValueError: invalid literal for int() with base 10: '0rc1'`: [Issue #57: Usage of pytest.__version__ breaks testing with prereleases - mirrors-countme - Pagure.io](https://pagure.io/mirrors-countme/issue/57)
- [x] [python-pytest-isort](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058523-python-pytest-isort/builder-live.log.gz): `E       ValueError: invalid literal for int() with base 10: '0rc1'`: [`get_pytest_version` version parsing breaks with prereleases · Issue #38 · stephrdev/pytest-isort](https://github.com/stephrdev/pytest-isort/issues/38)

## Possible pytest regressions

- [x] [python-html5lib](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058513-python-html5lib/builder-live.log.gz): Using `Config.rootdir` from `pytest_configure` (fix in progress in pytest: https://github.com/pytest-dev/pytest/pull/9400)
- [x] [python-promise](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058521-python-promise/builder-live.log.gz): Accesses `assert assert_exc.traceback[-1].path.strpath == __file__` (from `ExceptionInfo`), but `path` is a `PosixPath` now. Sounds like we falsely assumed that was internal only? PR open: https://github.com/syrusakbary/promise/pull/96; pytest issue: #9423
- [x] [python-pytest-randomly](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058526-python-pytest-randomly/builder-live.log.gz): `E       AttributeError: 'Module' object has no attribute 'instance'` - https://github.com/pytest-dev/pytest-randomly/pull/423 and https://github.com/pytest-dev/pytest/issues/9486
- [x] [python-pytest-bdd](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058520-python-pytest-bdd/builder-live.log.gz): pytester test fails because more tests than expected are `deselected` <br> ~~Bisected to 6d6cfd839aa850061ddef65922eddb7567280e5f. Was this just not tested before?~~ See https://github.com/pytest-dev/pytest-bdd/pull/470 and https://github.com/pytest-dev/pytest/issues/9471
- [x] `pytest-isort`: Fails with `if session is None: assert parent is None` in `nodes.py` after https://github.com/stephrdev/pytest-isort/pull/39 <br> See https://github.com/pytest-dev/pytest/issues/9487


## Unclear

- [ ] [ansible](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058507-ansible/builder-live.log.gz): No idea, most tests seem to fail due to a `sys.exit` apparently? Upstream: https://github.com/ansible/ansible/issues/76577. <br> Skipped for now, could not get tests to run.
- [ ] [ansible-core](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058505-ansible-core/builder-live.log.gz): Various failures with `TypeError: mock_NamedTemporaryFile() missing 1 required positional argument: 'mocker'`. <br> Skipped for now, could not get tests to run.
- [ ] [poetry](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058503-poetry/builder-live.log.gz): Reported upstream: https://github.com/python-poetry/poetry/issues/4901. <br> Skipped for now, can't seem to reproduce locally.
- [x] [python-interrogate](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058515-python-interrogate/builder-live.log.gz): `ConftestImportFailure`, can't see why. <br> **Bisected to 614f5394b50ac88b7aa4e97b43acb08aca383d95 but I don't understand.** [PR sent](https://github.com/econchick/interrogate/pull/99) - not pytest issue.
- [ ] [python-josepy](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058517-python-josepy/builder-live.log.gz): `E   ModuleNotFoundError: No module named 'josepy'` <br> Can't seem to reproduce locally, had some issues with pytest-flake8 but works otherwise (via tox).
- [x] [python-parameterized](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058522-python-parameterized/builder-live.log.gz): Unit test fails <br> Breaks for me locally even on pytest 6 in the same way, last reference in tox.ini is to pytest 3. Also see https://github.com/wolever/parameterized/issues/34.
- [x] [python-sqlalchemy1.3](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058529-python-sqlalchemy1.3/builder-live.log.gz): "No tests ran": https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3332
- [x] [python-sibyl](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058530-python-sybil/builder-live.log.gz): Some `AssertionError` inside `exec` <br> ~~Bisected to 92ba96b0612e6b06bb8f4ab05bd75481d2504806 but I don't understand why...~~ Fixed in https://github.com/pytest-dev/pytest/pull/9447

**Reply by hroncok** [reply]
Thanks!

**Reply by nicoddemus** [reply]
(@The-Compiler took the liberty of editing your post to add a few references 👍)

**Reply by bluetech** [reply]
For sqlalchemy see https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/3332. It seems like it's only applied to sqlalchemy 1.4, not 1.3.

**Reply by The-Compiler** [reply]
As for the `ExceptionInfo.traceback[...].path` one: The PR looks good, but it seems to me that we should retain and perhaps deprecate the `path` attribute properly, instead of (more or less silently) replacing it, despite it being public API)?

**Reply by bluetech** [reply]
We should add it to the legacypath plugin. Will do it when I get a chance.

**Reply by bluetech** [reply]
Oh it's *replaced*, that would be more difficult. Will take a look.

**Reply by The-Compiler** [reply]
I think we should start creating follow up issues for specific items as things can get lost in the comments here - I created one for the `ExeceptionInfo` thing: #9423.

Also opened a PR for httplib2's wrong `skip` usage: https://github.com/httplib2/httplib2/pull/215

I couldn't get Ansible's tests to run locally easily, but I still plan to take a look at the others marked as unclear.

**Reply by The-Compiler** [reply]
For the 10 unknown ones, I so far managed to bisect 3 to a commit in pytest. However, so far I don't quite understand what exactly broke and why. Probably going to stop for today now.

**Reply by hroncok** [reply]
I've proposed packaging fixes for python-box and python-vulture, they can be removed from this list.

**Reply by nicoddemus** [reply]
Do you have the links @hroncok ? I think it would be nice to add the links with their resolution (instead of removing them).

**Reply by hroncok** [reply]
I've added them to the Fedora list: https://src.fedoraproject.org/rpms/pytest/pull-request/25#comment-92228

- [bandit](https://src.fedoraproject.org/rpms/bandit/pull-request/2)
- [python-box](https://src.fedoraproject.org/rpms/python-box/pull-request/1)
- [python-vulture](https://src.fedoraproject.org/rpms/python-vulture/pull-request/1)

**Reply by nicoddemus** [reply]
Thanks! Added those to Florian's post. 👍

**Reply by hroncok** [reply]
I've also edited the headline to indicate the packaging issues are Fedora specific, not upstream packaging related.

**Reply by bluetech** [reply]
Sent a PR for `interrogate`: https://github.com/econchick/interrogate/pull/99 I'd say it's not a pytest issue/regression, there are probably some existing issues about better recovery from `monkeypatch` patching stdlib modules but it's nothing new except that we use pathlib more.

**Reply by bluetech** [reply]
See https://github.com/pytest-dev/pytest/pull/9440/files#r775133826 for `sybil`

**Comment by musicinmybrain** [comment]
I took a look at `python-pytest-bdd`. A nuance is that it seems that more tests than expected are deselected, not in the actual test suite, but in the test suites that run run as part of certain individual tests. In fact, I just added a manual BR on `python3dist(pytest-xdist)` so that now no top-level tests are skipped (instead of one).

As downstream maintainer, I don’t really know how to fix these failures with a reasonable amount of investigative effort, so guess I’ll report them upstream and skip the affected tests for now.

**Reply by musicinmybrain** [reply]
I have [reported](https://github.com/pytest-dev/pytest-bdd/issues/466) the `python-pytest-bdd` failures upstream and skipped the affected tests, so it now [builds in the COPR](https://copr.fedorainfracloud.org/coprs/churchyard/pytest-7/build/3061091/).

**Comment by kalekundert** [comment]
I fixed the bug in python-diff-cover: Bachmann1234/diff_cover#264

**Comment by The-Compiler** [comment]
FYI @hroncok I looked at some more:

- [x] [python-ratelimiter](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058528-python-ratelimiter/builder-live.log.gz): `E   pytest.PytestRemovedIn7Warning: pytest.collect.File was moved to pytest.File` - **probably not worth trying to fix, it's related to code for Python < 3.5 and no upstream commits in over 4 years despite open issues/PRs**
- [x] [pg_activity](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058504-pg_activity/builder-live.log.gz): `from _pytest.tmpdir import TempdirFactory` via pytest_postgresql
- [x] [python-pytest-cases](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058524-python-pytest-cases/builder-live.log.gz): `from _pytest.fixtures import scopes`: https://github.com/smarie/python-pytest-cases/issues/241 (no patch, fix seems a bit more involved)
- [x] [python-fiat](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058514-python-fiat/builder-live.log.gz): `from _pytest.python import scopes` via `pytest-cases`
- [x] [python-pytest-postgresql](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058525-python-pytest-postgresql/builder-live.log.gz): `from _pytest.tmpdir import TempdirFactory`: https://github.com/ClearcodeHQ/pytest-postgresql/issues/533 (trivial patch in comments, though a proper fix would do a bit more)

**Comment by The-Compiler** [comment]
# Remaining issues

Since almost all failures are dealt with and the post above got a bit unwieldy, I'm starting a new one for the remaining ones:

- [x] [python-pytest-rerunfailures](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058527-python-pytest-rerunfailures/builder-live.log.gz): `E       AttributeError: 'list' object has no attribute 'items'` in `for col, (finalizers, exc) in self.stack.items():` <br>Already fixed in https://github.com/pytest-dev/pytest-rerunfailures/pull/151, which was [released](https://github.com/pytest-dev/pytest-rerunfailures/blob/master/CHANGES.rst) in pytest-rerunfailures 10.0 back in May, but Fedora still seems to be at 9.1.1 (cc @hroncok)
- [x] `pytest-isort` fails with `assert parent is not None`: Unclear, asking for help in https://github.com/pytest-dev/pytest/issues/9487
- [x] `pytest-randomly`: Mainly worried about the undocumented `instance` attribute removal, https://github.com/pytest-dev/pytest/issues/9486
- [x] `josepy`: Still unclear why it [fails](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03128647-python-josepy/builder-live.log.gz) with a `ModuleNotFoundError` for the module under test on Fedora. I can't reproduce locally, so this might in fact be some kind of Fedora packaging issue (though I can't explain why it would work on pytest 6). **Giving up.**
- [ ] [ansible](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03128063-ansible/builder-live.log.gz): No idea, most tests seem to fail due to a `sys.exit` apparently? Upstream: https://github.com/ansible/ansible/issues/76577. <br> Skipped for now, could not get tests to run.
- [ ] [ansible-core](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03128069-ansible-core/builder-live.log.gz): Various failures with `TypeError: mock_NamedTemporaryFile() missing 1 required positional argument: 'mocker'`. <br> Skipped for now, could not get tests to run.
- [x] [poetry](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03128170-poetry/builder-live.log.gz): Reported upstream: https://github.com/python-poetry/poetry/issues/4901. <br>Can't reproduce even when trying without an internet connection. Note that I'm not entirely sure if that's really the issue - while there are warnings indicating network requests, the actual test failures might be from a different problem. Anyways, **giving up** on this one.
- [ ] [ansible-runner](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03128229-python-ansible-runner/builder-live.log.gz): New failure in second Fedora run only, might just be a flaky test.

Taking another look at Ansible and Poetry now - for the others (and possibly those too), I don't think there is much more we can do from pytest's side.

**Reply by musicinmybrain** [reply]
>     * [ ]  `josepy`: Still unclear why it [fails](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03128647-python-josepy/builder-live.log.gz) with a `ModuleNotFoundError` for the module under test on Fedora. I can't reproduce locally, so this might in fact be some kind of Fedora packaging issue (though I can't explain why it would work on pytest 6).

I just submitted a PR for this: https://src.fedoraproject.org/rpms/python-josepy/pull-request/1

I’m not sure why setting `PYTHONPATH` became necessary where it was not before, but it’s a trivial fix in the Fedora packaging for `josepy`.

Here’s the successful COPR build for the PR: https://copr.fedorainfracloud.org/coprs/churchyard/pytest-7/build/3132715/

**Reply by The-Compiler** [reply]
I'm afraid I'll give up with Ansible as well. I get dozens of failing tests with pytest 6, and at least some of the thousands of failures with pytest 7 seem to be wrong API usage. To make things worse, something in their testsuite seems to change node IDs so it's a pain to even find the responsible tests...

Spent way too much time on this one now. Time for weekend instead!

**Reply by hroncok** [reply]
> ansible-runner: New failure in second Fedora run only, might just be a flaky test.

I've submitted 10 builds with each pytest: https://copr.fedorainfracloud.org/coprs/churchyard/pytest-7/package/python-ansible-runner/ and https://copr.fedorainfracloud.org/coprs/churchyard/pytest-6/package/python-ansible-runner/

**Reply by hroncok** [reply]
> [python-pytest-rerunfailures](https://download.copr.fedorainfracloud.org/results/churchyard/pytest-7/fedora-rawhide-x86_64/03058527-python-pytest-rerunfailures/builder-live.log.gz): `E       AttributeError: 'list' object has no attribute 'items'` in `for col, (finalizers, exc) in self.stack.items():` <br>Already fixed in https://github.com/pytest-dev/pytest-rerunfailures/pull/151, which was [released](https://github.com/pytest-dev/pytest-rerunfailures/blob/master/CHANGES.rst) in pytest-rerunfailures 10.0 back in May, but Fedora still seems to be at 9.1.1 (cc @hroncok)

I've noted this in the python-pytest-rerunfailures update bugzilla and linked it to the pytest update bugzilla: https://bugzilla.redhat.com/show_bug.cgi?id=1965106#c3

**Reply by The-Compiler** [reply]
Looks like the ansible-runner test is indeed flaky: 2/10 fail with pytest 6 (same issue), 5/10 with pytest 7. Looks like there's [Clear test flake (test_password_prompt) by jladdjr · Pull Request #472 · ansible/ansible-runner](https://github.com/ansible/ansible-runner/pull/472) (should be already in with the tested version), but perhaps that's not a full fix.
